mempool/bench_test.go

```diff
@@ -148,3 +155,159 @@ func BenchmarkUpdateRemoteClient(b *testing.B) {
 		doUpdate(b, mp, int64(i), txs)
 	}
 }
+
+// Benchmarks the time it takes a blocking iterator to access all transactions
+// in the mempool.
+func BenchmarkBlockingIterator(b *testing.B) {
+	app := kvstore.NewInMemoryApplication()
+	cc := proxy.NewLocalClientCreator(app)
+	mp, cleanup := newMempoolWithApp(cc)
+	defer cleanup()
+
+	const numTxs = 1000
+	txs := addTxs(b, mp, 0, numTxs)
+	require.Equal(b, numTxs, len(txs))
+	require.Equal(b, numTxs, mp.Size())
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		b.StopTimer()
+		iter := NewBlockingIterator(context.TODO(), mp, b.Name())
+		b.StartTimer()
+
+		// Iterate until all txs in the mempool are accessed.
+		for c := 0; c < numTxs; c++ {
+			if entry := <-iter.WaitNextCh(); entry == nil {
+				continue
+			}
+		}
+	}
+}
+
+// Benchmarks the time it takes multiple concurrent blocking iterators to access
+// all transactions in the mempool while transactions are being concurrently
+// added.
+
+func BenchmarkBlockingIteratorsWhileAddingTxs(b *testing.B) {
+	app := kvstore.NewInMemoryApplication()
+	cc := proxy.NewLocalClientCreator(app)
+	mp, cleanup := newMempoolWithApp(cc)
+	defer cleanup()
+
+	const numTxs = 1000
+	const numIterators = 10
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		b.StopTimer()
+		mp.Flush()
+		wg := sync.WaitGroup{}
+		wg.Add(numIterators)
+		b.StartTimer()
+
+		// Concurrent iterators.
+		for j := 0; j < numIterators; j++ {
+			go func() {
+				defer wg.Done()
+				b.StopTimer()
```

Comment:
This looks like a data race to me. I would rewrite it to avoid any potential problems.

```go
                 wg := sync.WaitGroup{}
		wg.Add(numIterators)
		
		// Concurrent iterators.
		iters := make([]*BlockingIterator, numIterators)
		for j := 0; j < numIterators; j++ {
			iters[j] = NewBlockingIterator(context.TODO(), mp, strconv.Itoa(j))
		}
		
		b.StartTimer()
		
			go func(iter *BlockingIterator) {
				defer wg.Done()
```

Reply:
Thanks for pointing out @melekes. I didn't notice any data race but I see the risk of happening. 

Anyway, I decided to remove these two last test benchmarks that add and remove txs concurrently because they were too slow and difficult to benchmark correctly. For the moment, I added a simple benchmark with multiple concurrent iterators accessing all txs in the mempool, without any extra routine modifying the mempool concurrently.